### PR TITLE
[NFC] Move DroppedVariableStats to its own file and redesign it to be extensible.

### PR DESCRIPTION
**Pull request by rastogishubham**
Move DroppedVariableStats code to its own file and change the class to have an extensible design so that we can use it to add dropped statistics to MIR passes and the instruction selector.

### CODE REVIEW

**Comment by rastogishubham** [comment]
@OCHyams Done!

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:165]: nit: whitespace

File: llvm/lib/CodeGen/DroppedVariableStats.cpp
Comment (OCHyams) [inline llvm/lib/CodeGen/DroppedVariableStats.cpp:150]: nit: whitespace

File: llvm/lib/CodeGen/DroppedVariableStats.cpp
Comment (OCHyams) [inline llvm/lib/CodeGen/DroppedVariableStats.cpp:131]: nit: I can see this is matching the previous interface but I don't think these need to be `std::string`s. Could be worth changing these to `StringRef`s while you're here? I'll leave that up to you.

(could come later/separately if you do change it)

**Review by OCHyams — APPROVED** [review]
Thanks, LGTM

(edit: though it possibly depends on [this comment](https://github.com/llvm/llvm-project/pull/115566#pullrequestreview-2429040525))

**Review by adrian-prantl — APPROVED** [review]

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (adrian-prantl) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:62]: this is redundant

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (adrian-prantl) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:111]: wasDropped()?

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (adrian-prantl) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:163]: cleanup();

File: llvm/lib/CodeGen/DroppedVariableStats.cpp
Comment (adrian-prantl) [inline llvm/lib/CodeGen/DroppedVariableStats.cpp:163]: Why is this part of the virtual function? It seems like it wouldn't change.

File: llvm/unittests/IR/DroppedVariableStatsIRTest.cpp
Comment (adrian-prantl) [inline llvm/unittests/IR/DroppedVariableStatsIRTest.cpp:2]: going above 80 Colm limit here

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:28]: Can you please include what each of the fields represents in the comment, and nit: should this be "A unique key that represents a debug variable"?

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:85]: nit: #dbg_value -> variable or source variable or debug variable?

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:112]: Important to note in the comment that this also updates the "dropped" count IMO. Not sure how to reflect that in the function name too, but if it could be that would be nice.

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:128]: I think this comment and function name could be improved for clarity.

The comment makes it sound like this function is visiting every VarID, but its that this function is called with every VarID. Maybe this is just a hangover from a previous iteration, since it also looks like `DebugVariablesBeforeSet` isn't used in the virtual override (neither IR nor MIR in the next patch)?

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:135]: In both comment and function name debug intrinsic should be debug record or something more general. Debug record is better than intrinsic, since this debug intrinsics are a thing of the past :D

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [inline llvm/include/llvm/CodeGen/DroppedVariableStats.h:103]: `private`?

**Review by OCHyams — COMMENTED** [review]
There might be room for further code deduplication but the cost of the boilerplate might not be worth it, so I think this broadly LGTM, if @adrian-prantl is happy with this approach too. I've got some inline nits, mostly to do with comments and naming.

In future, please can you avoid squashing your changes when addressing reviewer feedback? It makes it more difficult to understand what has changed from the original patch (which means I find myself re-reviewing the whole thing rather than the changes). Especially with GitHub's tendency to swallow inline review comments. (It's fine to squish all your local commits between rounds of review ofc, and it's not a hard rule. There are times where it makes sense).

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (rastogishubham) [reply llvm/include/llvm/CodeGen/DroppedVariableStats.h:112]: `checkIfVarDroppedAndUpdateCount`?

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (rastogishubham) [reply llvm/include/llvm/CodeGen/DroppedVariableStats.h:128]: Oh you're right, this is my bad, yes originally the function would visit every VarID, but it doesn't anymore because of some small changes I made. Let me update this

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (OCHyams) [reply llvm/include/llvm/CodeGen/DroppedVariableStats.h:112]: How about `updateDroppedCount` (or `...Counter`)? I feel like the bool return value + comment makes the rest of the functionality clear? ymmv.

File: llvm/include/llvm/CodeGen/DroppedVariableStats.h
Comment (rastogishubham) [reply llvm/include/llvm/CodeGen/DroppedVariableStats.h:60]: Here however, we can assert to make sure the stacks are non-empty

**Review by OCHyams — APPROVED** [review]
LGTM - before committing can you please see if any of the fields in the base class can be made private instead of protected (I forgot to look at those while looking at the access specifiers for the member functions).

I'd also like to gently point out that you squashed your changes again (see my previous comment - this makes it more time consuming and difficult to review your updates).

**Comment by rastogishubham** [comment]
@OCHyams sorry about the squashing, I missed that in your previous comment. I will make sure I keep that in mind for the future

**Comment by rastogishubham** [comment]
Had to revert with e914d97327ce5887e51c2d650987f2f48eda85c1 because of buildbot failures

**Pull request merged**
